substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java

```diff
@@ -0,0 +1,186 @@
+     * Emit records for each function: PROC32 S_FRAMEPROC S_END and line number records. (later:
+     * type records as required).
+     *
+     * @param primaryEntry primary entry for this function
+     * @param methodName method name alias as it will be seen by the user
+     */
+    private void build(PrimaryEntry primaryEntry, String methodName) {
+        final Range primaryRange = primaryEntry.getPrimary();
+
+        /* S_PROC32 add function definition. */
+        int functionTypeIndex = addTypeRecords(primaryEntry);
+        byte funcFlags = 0;
+        CVSymbolSubrecord.CVSymbolGProc32Record proc32 = new CVSymbolSubrecord.CVSymbolGProc32Record(cvDebugInfo, methodName, 0, 0, 0, primaryRange.getHi() - primaryRange.getLo(), 0, 0,
+                        functionTypeIndex, primaryRange.getLo(), (short) 0, funcFlags);
+        addToSymbolRecord(proc32);
+
+        /* S_FRAMEPROC add frame definitions. */
+        int asynceh = 1 << 9; /* Async exception handling (vc++ uses 1, clang uses 0). */
+        int localBP = 1 << 14; /* Local base pointer = SP (0=none, 1=sp, 2=bp 3=r13). */
+        int paramBP = 1 << 16; /* Param base pointer = SP. */
+        int frameFlags = asynceh + localBP + paramBP; /* NB: LLVM uses 0x14000. */
+        addToSymbolRecord(new CVSymbolSubrecord.CVSymbolFrameProcRecord(cvDebugInfo, primaryRange.getHi() - primaryRange.getLo(), frameFlags));
+
+        /* TODO: add local variables, and their types. */
+        /* TODO: add block definitions. */
+
+        /* S_END add end record. */
+        addToSymbolRecord(new CVSymbolSubrecord.CVSymbolEndRecord(cvDebugInfo));
+        addLineNumberRecords(primaryEntry, methodName);
+    }
+
+    /**
+     * Rename function names for usability or functionality.
+     *
+     * First encountered main function becomes class.main (unless replaceMainFunctionName is
+     * non-null). This is for usability.
+     *
+     * All other functions become class.function.999 (where 999 is a hash of the arglist). This is
+     * because The standard link.exe can't handle odd characters (parentheses or commas, for
+     * example) in debug information, so
+     *
+     * This does not affect external symbols used by linker.
+     *
+     * TODO: strip illegal characters from arg lists instead ("link.exe" - safe names)
+     *
+     * @param range Range contained in the method of interest
+     * @return user debugger friendly method name
+     */
+    private String fixMethodName(Range range) {
+        final String methodName;
+        if (replaceMainFunctionName != null && noMainFound && range.getMethodName().equals("main")) {
+            noMainFound = false;
+            methodName = replaceMainFunctionName;
+        } else if (emitUnadornedMain && noMainFound && range.getMethodName().equals("main")) {
+            /* TODO: check for static void main(String args[]) instead of just main(...). */
+            noMainFound = false;
+            methodName = range.getClassAndMethodName();
```

Comment:
As far as usability goes, I think this is fine. However, I don’t think we need all these "options". Could we simplify this and just use the first `main` we find. This should only be temporary anyway (until the actual `main` becomes available via `DebugInfoProvider` as I have already suggested).

Reply:
For now, the first main function will be  'package.Class.main' no 'package.Class.main_9999999'.